cmd/orb.go

```diff
@@ -740,3 +758,144 @@ https://circleci.com/orbs/registry/orb/%s
+}
+
+type ExampleUsageSchema struct {
+	Version   string      `yaml:"version,omitempty"`
+	Orbs      interface{} `yaml:"orbs,omitempty"`
+	Jobs      interface{} `yaml:"jobs,omitempty"`
+	Workflows interface{} `yaml:"workflows"`
+}
+
+type ExampleSchema struct {
+	Description string             `yaml:"description,omitempty"`
+	Usage       ExampleUsageSchema `yaml:"usage,omitempty"`
+	Result      ExampleUsageSchema `yaml:"result,omitempty"`
+}
+
+func packOrb(opts orbOptions) error {
+	// Travel our Orb and build a tree from the YAML files.
+	// Non-YAML files will be ignored here.
+	_, err := os.Stat(filepath.Join(opts.args[0], "@orb.yml"))
+	if err != nil {
+		return errors.Wrap(err, "@orb.yml file not found, are you sure this is the Orb root?")
+	}
+
+	tree, err := filetree.NewTree(opts.args[0], "executors", "jobs", "commands", "examples")
+	if err != nil {
+		return errors.Wrap(err, "An error occurred trying to build the tree")
+	}
+
+	y, err := yaml.Marshal(&tree)
+	if err != nil {
+		return errors.Wrap(err, "Failed trying to marshal the tree to YAML")
+	}
+
+	// Create generic YAML node.
+	var node yaml.Node
+	err = yaml.Unmarshal(y, &node)
+	if err != nil {
+		return errors.Wrap(err, "Failed unmarshal YAML tree")
+	}
+
+	err = travelOrbTree(&node, opts.args[0])
+	if err != nil {
+		return errors.Wrap(err, "Failed trying to travel Orb YAML")
+	}
+
+	final, err := yaml.Marshal(&node)
+	if err != nil {
+		return errors.Wrap(err, "Failed trying to marshal Orb YAML")
+	}
+
+	// Unmarshal again into a struct so that it can be
+	// marhsalled again but with custom-ordered keys
+	var orb OrbSchema
+	err = yaml.Unmarshal(final, &orb)
+	if err != nil {
+		return errors.Wrap(err, "Failed unmarshal YAML tree")
+	}
+
+	finalOrdered, err := yaml.Marshal(&orb)
+	if err != nil {
+		return errors.Wrap(err, "Failed trying to marshal Orb YAML")
+	}
+
+	fmt.Println(string(finalOrdered))
+
+	return nil
+}
+
+// Travel down a YAML node, replacing values as we go.
+func travelOrbTree(node *yaml.Node, orbRoot string) error {
+	// View: https://regexr.com/582gb
+	includeRegex, err := regexp.Compile(`(?U)^<<\s*include\((.*\/*[^\/]+)\)\s*?>>$`)
+	if err != nil {
+		return err
+	}
+	// View: https://regexr.com/57hl9
+	maybeHeredocRegex, err := regexp.Compile(`(<<[\s\#\^\/]*[^<<]*)`)
+	if err != nil {
+		return err
+	}
+
+	// If we're dealing with a ScalarNode, we can replace the contents.
+	// Otherwise, we recurse into the children of the Node in search of
+	// a matching regex.
+	if node.Kind == yaml.ScalarNode && node.Value != "" {
+		includeMatches := includeRegex.FindStringSubmatch(node.Value)
+		if len(includeMatches) > 0 {
+			filepath := filepath.Join(orbRoot, includeMatches[1])
+			file, err := ioutil.ReadFile(filepath)
+			if err != nil {
+				return errors.Wrap(err, fmt.Sprintf("Could not open %s for inclusion in Orb", filepath))
+			}
+
+			node.Value = string(file)
+		}
+
+		// Escape unclosed `<<` for heredocs. First we partially match the syntax,
```

Comment:
I don't think this should be a feature of orb pack, as it's difficult to do right in all situations (what happens if a heredoc is already escaped here? what happens if I am doing something like `cat <<EOF >> appended-to.txt`?) and it means that copying config from a config into your orb will have sharp edges.

I think we should be really wary of making "orb config" (the input to orb packing) different from regular config (the output of orb packing), except in really obvious ways (like inserting scripts).

Reply:
I agree. This is something we added somewhat after the fact, but I'll look into pulling it out - would it be worth making a new PR to add this behaviour to both the config and orb pack (if we can refine the behaviour)? I think it's generally a positive addition, but hard to get right.

Reply:
I agree that it'd be positive if we could do it right, but it feels too magical to me, I think - it feels simpler to just always require escaping.

If we could do it, I think it'd be better to do it in config compilation itself (server side) so it'd work everywhere, including config being executed.